envs/monkey_zoo/blackbox/utils/gcp_machine_handlers.py

```diff
@@ -1,35 +1,47 @@
 import logging
+import os
 import subprocess
 
 LOGGER = logging.getLogger(__name__)
 
 
 class GCPHandler(object):
-
+    # gcloud commands
     AUTHENTICATION_COMMAND = "gcloud auth activate-service-account --key-file=%s"
     SET_PROPERTY_PROJECT = "gcloud config set project %s"
     MACHINE_STARTING_COMMAND = "gcloud compute instances start %s --zone=%s"
     MACHINE_STOPPING_COMMAND = "gcloud compute instances stop %s --zone=%s"
 
+    # Default configuration parameters
+    DEFAULT_RELATIVE_KEY_PATH = "../../gcp_keys/gcp_key.json"
+    DEFAULT_ZONE = "europe-west3-a"
+    DEFAULT_PROJECT = "guardicore-22050661"
+
     def __init__(
         self,
-        key_path="../gcp_keys/gcp_key.json",
-        zone="europe-west3-a",
-        project_id="guardicore-22050661",
+        relative_key_path=DEFAULT_RELATIVE_KEY_PATH,
+        zone=DEFAULT_ZONE,
+        project_id=DEFAULT_PROJECT,
     ):
         self.zone = zone
-        try:
-            # pass the key file to gcp
-            subprocess.call(GCPHandler.get_auth_command(key_path), shell=True)  # noqa: DUO116
-            LOGGER.info("GCP Handler passed key")
-            # set project
-            subprocess.call(  # noqa: DUO116
-                GCPHandler.get_set_project_command(project_id), shell=True
+        abs_key_path = GCPHandler.get_absolute_key_path(relative_key_path)
+        # pass the key file to gcp
+        subprocess.call(GCPHandler.get_auth_command(abs_key_path), shell=True)  # noqa: DUO116
+        LOGGER.info("GCP Handler passed key")
+        # set project
+        subprocess.call(GCPHandler.get_set_project_command(project_id), shell=True)  # noqa: DUO116
+        LOGGER.info("GCP Handler set project")
+        LOGGER.info("GCP Handler initialized successfully")
+
+    @staticmethod
+    def get_absolute_key_path(relative_key_path: str) -> str:
+        file_dir = os.path.dirname(os.path.realpath(__file__))
+        absolute_key_path = os.path.join(file_dir, relative_key_path)
```

Comment:
> Since you're just joining file_dir (which is absolute) and relative_key_path (which is, presumably, relative), the result is a relative path.
I disagree, the result is an absolute path, because it will be resolved the same way, regardless of your cwd. `/usr/var/infection_monkey` and `/usr/var/infection_monkey/utils/common/../../` are both absolute paths that resolve the same way.

> More importantly, though, is if you have a relative path to the key, can't you just call os.path.realpath(relative_key_path) directly? Why join relative_key_path with file_dir?
Because the output of `os.path.realpath(relative_key_path)` depends on your cwd. The file_dir is the same regardless of cwd and is resolved the same way.

> Finally, is an absolute path necessary?
No, but it's clear to the user of `get_absolute_key_path` what he's getting. If you get a relative path, you still wonder what's it relative to, how do I make sure this path resolves?

Reply:
I've normalized the path(I was unsure about readability, because this is basically a no-op) and added a comment.

Reply:
> > Since you're just joining file_dir (which is absolute) and relative_key_path (which is, presumably, relative), the result is a relative path.
> 
> I disagree, the result is an absolute path, because it will be resolved the same way, regardless of your cwd. `/usr/var/infection_monkey` and `/usr/var/infection_monkey/utils/common/../../` are both absolute paths that resolve the same way.

Python disagrees with you. Since the `..` specifies a directory that is relative to the rest of the path, passing it to `os.path.abspath()` results in a path that is fully resolved.

![image](https://user-images.githubusercontent.com/19957806/117958476-0d962580-b2e9-11eb-8d38-6c8b74f34962.png)

To make our Python code easy to understand, our definition of absolute path should be consistent with Python's.

> > More importantly, though, is if you have a relative path to the key, can't you just call os.path.realpath(relative_key_path) directly? Why join relative_key_path with file_dir?
> 
> Because the output of `os.path.realpath(relative_key_path)` depends on your cwd. The file_dir is the same regardless of cwd and is resolved the same way.
>
> > Finally, is an absolute path necessary?
> 
> No, but it's clear to the user of `get_absolute_key_path` what he's getting. If you get a relative path, you still wonder what's it relative to, how do I make sure this path resolves?

Makes sense.

> I've normalized the path(I was unsure about readability, because this is basically a no-op) and added a comment.

`realpath()` calls `normpath()`, resolves symlinks, and then calls `abspath()` [on Linux](https://github.com/python/cpython/blob/6574334a68aa324394a6fd1f855ecbad20432b1e/Lib/posixpath.py#L390-L395).  [On windows](https://github.com/python/cpython/blob/cb6f3d7163c611a7772da8969475e47fbdd147af/Lib/ntpath.py#L638-L686), the first thing it does is call `normpath()`, before ultimately calling [GetFinalPathNameByHandle](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfinalpathnamebyhandlea?redirectedfrom=MSDN.). I'd say just call `realpath()` at the end so you get something that is definitively an absolute path.